src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java

```diff
@@ -280,81 +350,65 @@ private static boolean isInLambda(DetailAST paramDef) {
      */
     private static DetailAST findClassOrConstructorOrMethodInWhichItIsDefined(DetailAST ast) {
         DetailAST astTraverse = ast;
-        while (!(astTraverse.getType() == TokenTypes.METHOD_DEF
-                || astTraverse.getType() == TokenTypes.CLASS_DEF
-                || astTraverse.getType() == TokenTypes.ENUM_DEF
-                || astTraverse.getType() == TokenTypes.CTOR_DEF)) {
+        while (astTraverse.getType() != TokenTypes.METHOD_DEF
+                && astTraverse.getType() != TokenTypes.CLASS_DEF
+                && astTraverse.getType() != TokenTypes.ENUM_DEF
+                && astTraverse.getType() != TokenTypes.CTOR_DEF) {
             astTraverse = astTraverse.getParent();
         }
         return astTraverse;
     }
 
     /**
-     * Check if both the Variable are same.
+     * Check if both the Variables are same.
      * @param ast1 Variable to compare
      * @param ast2 Variable to compare
-     * @return true if both the variable are same, otherwise false
+     * @return true if both the variables are same, otherwise false
      */
     private static boolean isSameVariables(DetailAST ast1, DetailAST ast2) {
         final DetailAST classOrMethodOfAst1 =
             findClassOrConstructorOrMethodInWhichItIsDefined(ast1);
         final DetailAST classOrMethodOfAst2 =
             findClassOrConstructorOrMethodInWhichItIsDefined(ast2);
-
-        final String identifierOfAst1 =
-            classOrMethodOfAst1.findFirstToken(TokenTypes.IDENT).getText();
-        final String identifierOfAst2 =
-            classOrMethodOfAst2.findFirstToken(TokenTypes.IDENT).getText();
-
-        return identifierOfAst1.equals(identifierOfAst2);
+        return classOrMethodOfAst1 == classOrMethodOfAst2;
     }
 
     /**
-     * Inserts a variable at the topmost scope stack.
-     * @param ast the variable to insert
+     * Check if both the variables are in the same loop.
+     * @param ast1 variable to compare.
+     * @param ast2 variable to compare.
+     * @return true if both the variables are in the same loop.
      */
-    private void insertVariable(DetailAST ast) {
-        final Map<String, DetailAST> state = scopeStack.peek();
-        final DetailAST astNode = ast.findFirstToken(TokenTypes.IDENT);
-        state.put(astNode.getText(), astNode);
+    private static boolean isInTheSameLoop(DetailAST ast1, DetailAST ast2) {
+        DetailAST loop1 = ast1.getParent();
+        while (loop1 != null && !isLoopAst(loop1.getType())) {
+            loop1 = loop1.getParent();
+        }
+        DetailAST loop2 = ast2.getParent();
+        while (loop2 != null && !isLoopAst(loop2.getType())) {
+            loop2 = loop2.getParent();
+        }
+        return loop1 == null && loop2 == null
+                || loop1 != null && loop1 == loop2;
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Use equals() to compare object references.](https://www.codacy.com/app/checkstyle/checkstyle/file/1111847767/issues/source?bid=2193633&fileBranchId=2892126#l392)